### ISSUE TITLE
Consistent NaN crashes

### ISSUE THREAD

**Issue by jsteggink** [deeplearning4j/deeplearning4j#1585]
Since version 3.9 I have been having trouble training my GravesLSTM network. It seems to be related to the amount of data. I have tried many mini batch sizes, all sorts of network configurations. With regularization, without, all sorts of learning rates, gradientNormalization, they have no influence, it will always crash at the same score iteration. Depending on the size of data the crashpoint will shift from the first to a subsequent epoch. The larger the training data, the further in the epochs it will crash. In total I have 120.000 rows, 92 input classes, 3 output classes, and the sequence length is 30. With these data it crashes in the second epoch.

Exception:
https://gist.github.com/jsteggink/89e7c36a877f2eb641c756bdc373521d

**Comment by jsteggink** [comment]
![image](https://cloud.githubusercontent.com/assets/978411/15456302/141fdf50-206c-11e6-8394-2f85b791dfb3.png)

**Comment by raver119** [comment]
It maybe some bug within HistogramListener itself. Have you tried to disable it, and just use ScoreIterationListener to see if that's the case?

But still, obviously HistogramListener shouldn't crash.

**Comment by jsteggink** [comment]
Without the HistogramIterationListener it's the same issue. I can run it without and send the exception if yout want, so you can narrow down the issue. However it will take a while, since I will have to start the training again.

**Comment by raver119** [comment]
Yes, do that please if that's possible.

**Comment by jsteggink** [comment]
Sorry, I forgot, no exception, just NaN at the same iteration it crashes with the HistogramIterationListener.

o.d.o.l.ScoreIterationListener - Score at iteration 1341 is 11.537639902156043
c.s.n.w.t.WordbreakerTrainingIterator - batch: 143 numCharacters: 92 maxWordLength: 30
o.d.o.l.ScoreIterationListener - Score at iteration 1342 is 11.06500806641667
c.s.n.w.t.WordbreakerTrainingIterator - batch: 144 numCharacters: 92 maxWordLength: 30
o.d.o.l.ScoreIterationListener - Score at iteration 1343 is 11.03507276386018
c.s.n.w.t.WordbreakerTrainingIterator - batch: 145 numCharacters: 92 maxWordLength: 30
o.d.o.l.ScoreIterationListener - Score at iteration 1344 is NaN
c.s.n.w.t.WordbreakerTrainingIterator - batch: 146 numCharacters: 92 maxWordLength: 30
o.d.o.l.ScoreIterationListener - Score at iteration 1345 is NaN
c.s.n.w.t.WordbreakerTrainingIterator - batch: 147 numCharacters: 92 maxWordLength: 30
o.d.o.l.ScoreIterationListener - Score at iteration 1346 is NaN
c.s.n.w.t.WordbreakerTrainingIterator - batch: 148 numCharacters: 92 maxWordLength: 30
o.d.o.l.ScoreIterationListener - Score at iteration 1347 is NaN

**Comment by AlexDBlack** [comment]
Check your input data and labels at that iteration.
Also: More UI output, just before it starts with the NaN would be good... parameter histograms and gradient mean magnitudes for each layer would be good to help diagnose this.

**Comment by saudet** [comment]
Might be related to https://github.com/deeplearning4j/libnd4j/issues/191

**Comment by jsteggink** [comment]
I don't think it is the data itself. The data has a random order every time I initialise the data set iterator, it happens in the the third epoch, so it already has seen all the data twice.

I debugged the training and it seems it all goes wrong here
![image](https://cloud.githubusercontent.com/assets/978411/15467981/b81b5f88-20e1-11e6-9a2f-17bd27160510.png)

**Comment by jsteggink** [comment]
I'm not exactly sure at what point it happens, but just before I get NaN scores I have this gradient:
`{0_W=[[0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [   �,    �,    �,    �],
 [   �,    �,    �,    �],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [   �,    �,    �,    �],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, -0,00, -0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [-0,00, -0,00, 0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, -0,00, 0,00, 0,00],
 [-0,00, -0,00, 0,00, 0,00],
 [0,00, 0,00, 0,00, -0,00],
 [-0,00, 0,00, 0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, -0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, -0,00, 0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, -0,00, -0,00, 0,00],
 [-0,00, -0,00, 0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, -0,00, 0,00, 0,00],
 [-0,00, -0,00, 0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, -0,00, 0,00],
 [0,00, 0,00, -0,00, -0,00],
 [-0,00, -0,00, 0,00, -0,00],
 [0,00, -0,00, -0,00, -0,00],
 [0,00, 0,00, -0,00, -0,00],
 [0,00, 0,00, -0,00, -0,00],
 [0,00, -0,00, 0,00, 0,00],
 [-0,00, 0,00, -0,00, 0,00],
 [0,00, -0,00, 0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, 0,00, -0,00, 0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, 0,00, -0,00],
 [0,00, 0,00, 0,00, 0,00],
 [0,00, -0,00, 0,00, -0,00],
 [0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, -0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, 0,00, 0,00, -0,00],
 [-0,00, 0,00, 0,00, 0,00],
 [0,00, 0,00, -0,00, 0,00],
 [0,00, -0,00, -0,00, 0,00],
 [-0,00, 0,00, 0,00, -0,00],
 [0,00, 0,00, -0,00, 0,00],
 [0,00, 0,00, -0,00, 0,00],
 [-0,00, 0,00, 0,00, -0,00],
 [-0,00, -0,00, -0,00, 0,00],
 [0,00, -0,00, -0,00, 0,00],
 [-0,00, 0,00, -0,00, -0,00],
 [0,00, 0,00, 0,00, -0,00],
 [-0,00, 0,00, 0,00, -0,00],
 [-0,00, -0,00, 0,00, -0,00],
 [0,00, -0,00, -0,00, 0,00],
 [0,00, 0,00, 0,00, -0,00],
 [0,00, 0,00, 0,00, 0,00],
 [-0,00, 0,00, -0,00, 0,00],
 [-0,00, -0,00, -0,00, 0,00],
 [0,00, -0,00, 0,00, 0,00]], 0_RW=[-0,00, 0,00, 0,00, 0,00, 0,00, 0,00, 0,00], 0_b=[0,00, -0,00, -0,00, -0,00], 1_W=[-0,00, 0,00, 0,00, 0,00], 1_RW=[-0,00, 0,00, 0,00, 0,00, 0,00, 0,00, 0,00], 1_b=[0,00, -0,00, -0,00, -0,00], 2_W=[0,00, -0,00, -0,00], 2_b=[-0,00, -0,00, 0,00]}`

**Comment by eraly** [comment]
@jsteggink What os are you on? And this is with nd4j-native or cuda?

**Comment by jsteggink** [comment]
This is on Windows with nd4j-native

**Comment by eraly** [comment]
Thanks. I am on a mac and linux. I can run on my end and see what happens. But if I can't reproduce the issue the other guys (who have windows) will look into it :)

**Comment by jsteggink** [comment]
@eraly I just ran the code on Linux (Ubuntu 14.04) and I have no problem at all. So I guess it's Windows related. Don't know why I hadn't tried that before.

**Comment by eraly** [comment]
@jsteggink But I get the feeling we will still want to figure out what the problem was.

**Comment by jsteggink** [comment]
I agree, something libnd4j related probably?

**Comment by eraly** [comment]
Eh. I can't really tell yet. I am running through on my mac fine so far. Raver will probably try to reproduce it on his laptop.

**Comment by DevinCalado** [comment]
Hey guys.  I get a similar issue with the � symbols appearing in my training data causing the same NaN issue.  The �'s only appear in the data set AFTER i call dataset.normalizeZeroMeanZeroUnitVariance()

Tehre are no � symbols in my data if I don't run dataset.normalizeZeroMeanZeroUnitVariance()

**Comment by eraly** [comment]
@DevinCalado You should use standard scaler instead. 
import org.nd4j.linalg.dataset.api.iterator.StandardScaler;
StandardScaler scaler = new StandardScaler();
scaler.fit(iterator);

**Comment by DevinCalado** [comment]
@eraly I don't understand why the standard scalar will work instead but I'll give it a try when I get home. Maybe I don't fully understand why normalizeZeroMeanZeroUnitVariance is giving bad symbols.    I'll post an update after I try what you suggested. Cheers!

**Comment by eraly** [comment]
@DevinCalado Basically because it is a method we are going to deprecate. You have to do a transform after the fit on every batch too. If it doesn't work file a separate issue.

**Comment by DevinCalado** [comment]
@eraly Okay thanks for the advice. I'll follow up if it is a separate issue.

**Comment by eraly** [comment]
Resolved. Closing.
